nifi-nar-bundles/nifi-cdc/nifi-cdc-postgresql-bundle/nifi-cdc-postgresql-processors/src/main/java/org/apache/nifi/cdc/postgresql/event/Reader.java

```diff
@@ -0,0 +1,242 @@
+        try {
+            // Loading data types.
+            if (replicationDecoder.isDataTypesEmpty()) {
+                replicationDecoder.loadDataTypes(queryConn);
+            }
+        } catch (SQLException e) {
+            throw new SQLException("Failed to load data types", e);
+        }
+
+        this.includeBeginCommit = includeBeginCommit;
+        this.includeAllMetadata = includeAllMetadata;
+    }
+
+    /**
+     * Constructor only for unit tests.
+     *
+     * @see org.apache.nifi.cdc.postgresql.event.MockReader
+     */
+    public Reader(Long lsn, boolean includeBeginCommit, boolean includeAllMetadata) {
+    }
+
+    /**
+     * Reads a pending event from the slot replication since the last
+     * feedback. If received buffer is null, there is not more pending events. The
+     * binary event buffer is decoded and converted to a JSON String.
+     *
+     * @return HashMap (String, Object)
+     * @throws SQLException
+     *                      if read event failed
+     * @throws IOException
+     *                      if decode event failed
+     *
+     * @see org.apache.nifi.cdc.postgresql.event.Slot
+     * @see org.apache.nifi.cdc.postgresql.event.Decoder
+     */
+    public HashMap<String, Object> readMessage() throws SQLException, IOException {
+        ByteBuffer buffer;
+        try {
+            buffer = this.replicationStream.readPending();
+        } catch (Exception e) {
+            throw new SQLException("Failed to read pending events", e);
+        }
+
+        if (buffer == null) {
+            return null;
+        }
+
+        // Binary buffer decode process.
+        HashMap<String, Object> message;
+        try {
+            message = replicationDecoder.decodeLogicalReplicationBuffer(buffer, this.includeBeginCommit,
+                    this.includeAllMetadata);
+        } catch (Exception e) {
+            throw new IOException("Failed to decode event buffer", e);
+        }
+
+        // Messages not requested/included.
+        if (message.isEmpty()) {
+            return message;
+        }
+
+        LogSequenceNumber lsn = this.replicationStream.getLastReceiveLSN();
+        Long lsnLong = lsn.asLong();
+
+        // Some messagens don't have LSN (e.g. Relation).
+        if (lsnLong > 0) {
+            message.put("lsn", lsnLong);
+        }
+
+        return message;
+    }
+
+    /**
+     * Converts event message to JSON String.
+     *
+     * @param message
+     *                The key-value message.
+     * @return String
+     * @throws IOException
+     *                     if convert message failed
+     */
+    public String convertMessageToJSON(HashMap<String, Object> message) throws IOException {
+        ObjectMapper jsonMapper = new ObjectMapper();
+        try {
+            return jsonMapper.writeValueAsString(message);
+        } catch (JsonProcessingException e) {
+            throw new IOException("Failed to convert event message to JSON", e);
+        }
+    }
```

Comment:
Recommend moving this utility method to another class since it is specific to JSON conversion.